cpp/src/arrow/util/align_util.cc

```diff
@@ -94,12 +135,17 @@ bool CheckAlignment(const Table& table, int64_t alignment,
   return all_aligned;
 }
 
+// Most allocators require a minimum of 8-byte alignment.
+constexpr int64_t kMinimumAlignment = 8;
```

Comment:
The problem arises when we are checking for value alignment.  For example, if we encounter a buffer that requires 4 byte alignment then we will call:

```
EnsureAlignment(buffer, 4, memory_pool);
```

If that buffer was not 4-byte aligned then we would try and allocate a new buffer:

```
    ARROW_ASSIGN_OR_RAISE(
        auto new_buffer,
        AllocateBuffer(buffer->size(), 4, memory_pool));
```

This would then raise.

Another option would be to change the signature of `EnsureAlignment`:

```
Result<std::shared_ptr<Buffer>> EnsureAlignment(std::shared_ptr<Buffer> buffer,
                                                int64_t expected_alignment,
                                                int64_t new_alignment_if_needed,
                                                MemoryPool* memory_pool)
```

Then we could pass `4` for `expected_alignment` and `kDefaultBufferAlignment` for `new_alignment_if_needed`.  I wasn't sure if it was worth cluttering the signature of `EnsureAlignment` or not.

Reply:
> This would then raise.

Hmm, why would it? I think I'm missing something here.

Reply:
Hmm, we should fix our wrapper around `posix_memalign` then, IMHO.
I think we should also use `max(kDefaultBufferAlignment, alignment)` in this PR.

Reply:
>  I think we should also use max(kDefaultBufferAlignment, alignment) in this PR.

I think that is reasonable.  I'll update the comments to explain this.

Reply:
I've done this, and gotten rid of `kMinimumAlignment`.

> Hmm, we should fix our wrapper around posix_memalign then, IMHO.

We can do this in a follow-up PR.  I've created https://github.com/apache/arrow/issues/35793